README.rst
.. default-role:: code

.. image:: https://circleci.com/gh/noisycomputation/pymssql-linux/tree/master.svg?style=shield
        :target: https://circleci.com/gh/noisycomputation/pymssql-linux

A simple database interface for `Python`_ that builds on top of `FreeTDS`_ to
provide a Python DB-API (`PEP-249`_) interface to `Microsoft SQL Server`_.

.. _Microsoft SQL Server: http://www.microsoft.com/sqlserver/
.. _Python: http://www.python.org/
.. _PEP-249: http://www.python.org/dev/peps/pep-0249/
.. _FreeTDS: http://www.freetds.org/

The Original Pymssql Project Has Been Discontinued
==================================================

This repository has been forked from upstream pymssql
and modified to work with both Python 3.7 and 3.8,
whilst continuing to build successfully for
2.7 for a short while, for sentimental but quite
fundamental reasons.

There is no interest whatsoever in making this work
in Windows.

The README from the upstream repository is available at
`README_upstream.rst`.

Why Support Python 2 until its dying day?
-----------------------------------------

pymssql is most useful for interfacing with SQL Server 2005,
and if the target environment demands the usage of such old
software, it might also limit one's ability to run Python 3.

It also just felt wrong to drop Python 2 on account of its
deprecation when SQL Server 2005 is so long in the tooth.

Why Not Support Windows?
------------------------

Folks, upstream pymssql was taken out back and shot because
Microsoft released native drivers into the open source. Usage
with SQL Server 2005 requires version 11.0 or earlier
of the MS driver, which is quite tricky to compile and run on a
modern \*nix system due to an entire butterfly network of
dependencies on long-deprecated system libraries. If you've ever
tried to compile something non-trivial on Linux during the
mid-90s, you'll know the pain of which I speak.

If Microsoft had never honored us by bestowing upon us this gift
of deprecated code that hadn't been maintained for a decade,
the upstream pymssql would have been the only game in town and
would have happily chugged along. But Microsoft did honor us,
which led to the behind-the-courthouse death of pymssql, which
led to the earthly hell of trying to compile Microsoft-gathered
\*nix code that relied on libraries from the early 2000s.

I have zero interest in figuring out how to build pymssql in
a Windows CI. If someone wants to work with me to port
pymssql to the Commodore 64, the Amiga 500, or Windows 2.1,
let us waste no time. But forget Windows >= 3.1.

It must be childishly easy to get the official drivers working
on a Windows box anyways.

Build Instructions
------------------

One of the reasons the upstream repository became such a tangle
of inconsistent and outdated information is that it contained
a large amount of build-related cruft that accumulated over
the years. And that cruft was in addition to a README that
contained build instructions that *maybe* worked two versions
ago.

To avoid this fate, this repository will maintain only those
CI configs that are being used to build, test, and deploy
this package to PyPi. Currently, that universe consists of
one CI, *circleci*, chosen only because it was the repository of
the most valid build instructions at the time this repo
was forked from upstream. This may change in the future.

Though this README may provide some additional context,
the only set of build instructions that are guaranteed
to be valid will be contained in the config file(s) for
the current CI system being used. For the branch to which
this README document pertains, the CI system is circleci
and the build instructions are at `.circleci/config.yml`

Though referencing a CI configuration file is not as
user-friendly as writing a soliloquy, it has the benefit
of being verifiably accurate. If the circleci status badge
at the top of this README indicates that the last build
succeeded, then the circleci build instructions are valid.

At a high level, official builds are done using manylinux2010
Docker containers. See the following files for details:

- `.circleci/config.yml`
- `docker-compose.yml`
- `dev/build_manylinux_wheels.sh`

Tests
-----

The testing suite has been inherited from upstream. While
it shows as passing in circleci, it actually fails with
multiple errors, and the successful result in circleci
appears to be the result of the tests being passed into
circleci as shell scripts, whose error-ful exit codes
are not being caught by circleci.

Bringing this project under test will be left to a future
maintainer, hopefully a future maintainer of the upstream
package. This maintainer tested the resulting packages
in the maintainer's use case, and the packages have passed.

Yes, that is anecdotal.

Since the tests do not work, they have been disabled in
`.circleci/config.yml` by commenting out.



ChangeLog
Change Log
==========

Version 2.1.6 - 2020-06-05 (pymssql-linux)
=================================================

- Fix circleci build script

- Update docker-compose.yml to manylinux2010 images

- Prepare package for upload to PyPi as pymssql-linux

Version 2.1.5 - 2019-11-27 (pymssql-linux)
=================================================

- Add Python 3.8 support.

- Drop support for Python < 3.6

Version 2.1.4 - To be released
==============================

General
-------

- Allow linkage against FreeTDS (by dropping usage of deprecated
  ``DBVERSION_80`` symbol.) (GH-432)

- Stop using 7.1 as default value for the TDS protocol version used in connections.

  This is a backward incompatible change and affects connections using both
  `pymssql` and `_mssql`.

  Now you need to specify a TDS protocol version explicitly by using any of
  the supported mechanisms (in descending order of precedence):

  * Using the ``tds_version`` paramenter of ``pymssql.connect()`` and
    ``_mssql.connect()``
  * A ``TDSVER`` enviromnent variable (see FreeTDS documentation)
  * A ``freetds.conf`` file (see FreeTDS documentation)

- Drop support for versions of FreeTDS older than 0.91.

- Accept 7.3 as TDS protocol version when establishing a connection.

- Add Python 3.7 support

- Drop Python 3.3 support

Features
--------

- Support for new in SQL Server 2008 ``DATE``, ``TIME`` and ``DATETIME2`` data
  types (GH-156). The following conditions need to be additionally met so
  values of these column types can be returned from the database as their
  native corresponding Python data types instead of as strings:

  * Underlying FreeTDS must be 0.95 or newer.
  * TDS protocol version in use must be 7.3 or newer.

  Thanks Ed Avis for the implementation. (GH-331)

Bug fixes
---------

- Finish implementation of TDS protocol version 7.3 support by actually
  accepting ``"7.3"`` as TDS protocol version when establishing a connection.
  (GH-455)
- Fixed and expanded Python data types that can be used to pass ``VARBINARY``,
  ``BINARY`` and ``IMAGE`` SQL Server Stored Procedures input parameters.
  (GH-425). Thanks Bill Adams for the fix.
- Fix ``tds_version``  ``_mssql`` connection property value for TDS version.
  7.1 is actually 7.1 and not 8.0.

Version 2.1.3 - 2016-06-22 - Ramiro Morales
===========================================

General
-------

- Windows official binaries: Rollback changes to Windows binaries we had
  implemented in pymssql 2.1.2; go back to using:

  * A statically linked version of FreeTDS (v0.95.95)
  * No SSL support

- Update bundled Linux static version of FreeTDS to v0.95.95.

Features
--------

- We now publish Linux PEP 513 manylinux wheels on PyPI.

Bug fixes
---------

- Add support for reporting TDS version 7.3 is in use via the ``tds_version``
  property of a ``_mssql``-level connection.

Version 2.1.2 - 2016-02-10 - Ramiro Morales
===========================================

.. attention:: Windows users: You need to download and install additional DLLs

    pymssql version 2.1.2 includes a change in the official Windows binaries:
    FreeTDS isn't statically linked as it happened up to release 2.1.1, as that
    FreeTDS copy lacked SSL support.

    Please see http://pymssql.org/en/latest/freetds.html#windows for futher
    details.

    We are trying to find a balance between security and convenience and will
    be evaluating the situation for future releases. Your feedback is greatly
    welcome.

General
-------

- Drop support for Python 2.6.

- Add support for Python 3.5.

- Update shipped Linux FreeTDS static libs to 0.95

Features
--------

- Add ability to set TDS protocol version from pymssql when connecting to SQL
  Server. For the remaining pymssql 2.1.x releases its default value will be 7.1
  (GH-323)

- Add Dockerfile and a Docker image and instructions on how to use it (GH-258).
  This could be a convenient way to use pymssql without having to build stuff.
  See http://pymssql.readthedocs.org/en/latest/intro.html#docker
  Thanks Marc Abramowitz.

- Floating point values are now accepted as Stored Procedure arguments
  (GH-287). Thanks Runzhou Li (Leo) for the report and Bill Adams for the
  implementation.

- Send pymssql version in the appname TDS protocol login record field when the
  application doesn't provide one (GH-354)

Bug fixes
---------

- Fix a couple of very common causes of segmentation faults in presence of
  network a partition between a pymssql-based app and SQL Server (GH-147,
  GH-271) Thanks Marc Abramowitz. See also GH-373.

- Fix failures and inconsistencies in query parameter interpolation when
  UTF-8-encoded literals are present (GH-185). Thanks Bill Adams. Also, GH-291.

- Fix ``login_timeout`` parameter of ``pymssql.connect()`` (GH-318)

- Fixed some cases of ``cursor.rowcont`` having a -1 value after iterating
  over the value returned by pymssql cursor ``fetchmany()`` and ``fetchone()``
  methods (GH-141)

- Remove automatic treatment of string literals passed in queries that start
  with ``'0x'`` as hexadecimal values (GH-286)

- Fix build fatal error when using Cython >= 0.22 (GH-311)

Documentation
-------------

- Add installation instructions. Thanks Marc Abramowitz.

- Document DB-API-mandated exceptions.

- Enhance ``_mssql.MSSQLStoredProcedure.bind()`` docs.

- Enhance description of Azure connections requirements.

Internals
---------

- Add Appveyor hosted CI setup for running tests on Windows (GH-347)

- Travis CI: Use newer, faster, container-based infrastructure. Also, test
  against more than one FreeTDS version.

- Make it possible to build official release files (sdist, wheels) on Travis &
  AppVeyor.

Version 2.1.1 - 2014-11-25 - Ramiro Morales
===========================================

Features
--------

- Custom message handlers (GH-139)

  The DB-Library API includes a callback mechanism so applications can provide
  functions known as *message handlers* that get passed informative messages

dev/test_manylinux_wheels.sh
#!/bin/bash
# Need to test the wheels that where build with build_manylinux_wheels.sh

# Rename test template file for CI
cp /io/tests/tests.cfg.tpl /io/tests/tests.cfg

# Setup xml results folder for CI if it does not exist
if [ ! -d /io/results ]; then
    mkdir /io/results
fi


# Install Python dependencies and compile wheels
for PYBIN in /opt/python/*/bin; do
    "${PYBIN}/pip" install pytest SQLAlchemy Sphinx sphinx-rtd-theme Cython wheel
done

# Install the wheels that were built. Need to be able to connect to mssql and to run the pytest suite after install
for PYBIN in /opt/python/*/bin/; do
    "${PYBIN}/pip" install pymssql --no-index -f /io/dist
    "${PYBIN}/python" -c "import pymssql; pymssql.__version__;"
    export TEST_PY="$(${PYBIN}/python -c 'import platform; major, minor, patch = platform.python_version_tuple(); print(major+minor+patch)')"
    "${PYBIN}pytest" /io --junitxml=/io/results/${TEST_PY}_test_results.xml
done

rm /io/tests/tests.cfg


## Run SQL Alchemy Tests
#pushd /io
#for PYBIN in /opt/python/*/bin/; do
#    PYV=`python -c "import sys;t='{v[0]}'.format(v=list(sys.version_info[:1]));sys.stdout.write(t)";`
#    if [ $PYV == '2' ]; then
#        pip install mock
#    fi
#    "${PYBIN}/pip" install nose
#    "${PYBIN}python" tests/run_sqlalchemy_tests.py
#done
#popd


dev/build_manylinux_wheels.sh
#!/bin/bash
# Need to build linux wheel using manylinux
#
# https://github.com/pypa/manylinux
# https://github.com/pypa/python-manylinux-demo
# https://www.python.org/dev/peps/pep-0513/
#
# You may need to make the build script executable
# chmod +x /dev/build_wheels.sh
#
# Standard manylinux docker containers provided by pypa. For more information see the links above.
# docker pull quay.io/pypa/manylinux1_x86_64
# docker pull quay.io/pypa/manylinux1_i686
#
# The next set of instructions will let you run the container interactively
# Provide a container name so it is easier to reference later
# sudo docker run --name manylinux_x86_x64 -it -d --rm -v `pwd`:/io quay.io/pypa/manylinux1_x86_64
# docker ps
#
# Use the docker exec command to interact with our container
# docker exec -it manylinux_x86_x64 ls
# docker exec -it manylinux_x86_x64 ./io/dev/build_wheels.sh
#
# Stop the conatiner when done
# docker stop manylinux_x86_x64
#
# These docker commands will run, build the wheel, attempt to install and then finally upload
# docker run --name manylinux_x86_x64 --rm -v `pwd`:/io quay.io/pypa/manylinux1_x86_64 /io/dev/build_wheels.sh
# docker run --name manylinux_i686 --rm -v `pwd`:/io quay.io/pypa/manylinux1_i686 /io/dev/build_wheels.sh
#
# https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html
set -e -x

# Install freetds and use in build. Yum channel shows version 0.91. Retrieving latest stable release for builds.
export PYMSSQL_BUILD_WITH_BUNDLED_FREETDS=1

rm -rf /io/freetds
mkdir /io/freetds
curl -sS ftp://ftp.freetds.org/pub/freetds/stable/freetds-1.1.tar.gz > freetds.tar.gz
tar -zxvf freetds.tar.gz -C /io/freetds --strip-components=1

export CFLAGS="-fPIC"  # for the 64 bits version

# omit --with-openssl=no if freetds version <1.1
pushd /io/freetds
./configure --enable-msdblib \
  --prefix=/usr --sysconfdir=/etc/freetds --with-tdsver=7.1 \
  --disable-apps --disable-server --disable-pool --disable-odbc \
  --with-openssl=no --with-gnutls=no

make install
popd

# Make wheelhouse directory
rm -rf /io/wheelhouse
mkdir /io/wheelhouse

# Install Python dependencies and compile wheels for Python 3.7 and 3.8 only
for PYBIN in /opt/python/*/bin; do
    "${PYBIN}/pip" install --upgrade pip setuptools
    "${PYBIN}/pip" install pytest SQLAlchemy Sphinx sphinx-rtd-theme Cython wheel
    "${PYBIN}/pip" wheel /io/ -w /io/wheelhouse/
done

# Verify the wheels and move from *-linux_* to -manylinux_*
for whl in /io/wheelhouse/*.whl; do
   auditwheel repair "$whl" -w /io/wheelhouse/
done

# Remove non-manylinux wheels
find /io/wheelhouse/ -type f ! -name '*manylinux*' -delete

# Create .tar.gz dist if it doesn't exists.
if [ ! -f /io/dist/*tar.gz* ]; then
    [ ! -e /io/dist ] && mkdir /io/dist
    pushd /io/
    /opt/python/cp37-cp37m/bin/python setup.py sdist
    popd
fi

# Move wheels to dist for install and upload
mv /io/wheelhouse/* /io/dist/

# Remove wheel and egg directory for next container build (i686 vs x86_x64)
rm -rf /io/wheelhouse/ /io/.eggs/ /io/pymssql.egg-info/

# Cleanup FreeTDS directories
rm -rf /io/freetds/ # /io/misc/ /io/include/ /io/doc/ /io/samples/ /io/vms/ /io/wins32/

echo "Done building wheels."
